### ISSUE TITLE
Support removing the brackets with `<C-h>`, `<C-w>` and `<C-u>`

### ISSUE THREAD

**Issue by jakubkaczor** [echasnovski/mini.nvim#246]
### Contributing guidelines

- [X] I have read [CONTRIBUTING.md](https://github.com/echasnovski/mini.nvim/blob/main/CONTRIBUTING.md)
- [X] I have read [CODE_OF_CONDUCT.md](https://github.com/echasnovski/mini.nvim/blob/main/CODE_OF_CONDUCT.md)

### Module(s)

pairs

### Description

Pairs supports registering a pair to `<BS>`. The `<C-h>` has the same functionality, but the plugin doesn't acknowledge for it. Similarly with `<C-w>` and `<C-u>`. I would like to see analogous behavior with regard to the other, mentioned keys.

**Comment by echasnovski** [comment]
Thanks for suggestion!

I understand the reasoning behind `<C-h>` in Insert mode. Does adding explicit mapping like with `vim.keymap.set('i', '<C-h>', '<BS>', { remap = true })` solve this issue?

I don't understand how you expect other mappings to work. `<C-w>` deletes word before cursor and `<C-u>` deletes all characters before cursor. What are examples of how they should work?

**Comment by jakubkaczor** [comment]
@echasnovski, thanks for a quick response! Right, as the scope of other key maps is different, perhaps it needs further explanation. I will try to give examples of behavior and then, try to state it in general. The examples will be in the following form. I will present the state of buffer before pressing the keys, and after. May `|` denote cursor in insert mode.

# `<C-w>`
Before:
```
function(|)
```
After:
```
function|
```
The similar behavior occurs in plain NeoVim, with regard to the opening bracket. In general, when the cursor is after the opening brackets, the result of pressing `<C-w>` should be the same as that of `<BS>`. There is one case in which I have doubts. That is when the opening bracket isn't really a bracket. For example, the letter `a`. I would expect the following with `f` as an opening and closing bracket.
Before:
```
snif|fing
```
After:
```
|fing
```
So in this case the behavior would be different, then that of `<BS>`.

I would like to mention, that both of these behaviors are consistent with NeoVim. I believe NeoVim treats such special characters as “(” as words, despite the spacing.

# `<C-u>`
Before:
```
This is an example (|) sentence.
```
After:
```
| sentence.
```
This seems to be more straightforward, then `<C-w>`. Simply delete all the characters before the cursor, and the closing bracket.

# '<C-h>`
With regard to the mapping of `<C-h>` to `<BS>`, I tried to do this a long time ago with some other plugin, and if I remember correctly, it resulted in unwanted side effect. I will try to dig it out, perhaps I mentioned the reason in some discussion.

**Comment by echasnovski** [comment]
Thanks, I understand it better now.

This is an interesting idea. I'll rephrase it in terms that are more aligned with 'mini.pairs'. For both `<C-w>` and `<C-u>` act like `<BS>` is acting now:
- If cursor is inside registered pair (like `aa (|) bb`), add additional `<Del>`. So for `<C-w>`, pressing it would be equivalent to typing `<C-w><Del>` without 'mini.pairs'. Same for `<C-u>`.
- Otherwise, act like plain `<C-w>` or `<C-u>`.

If this is what you would expect, then I think it should be achievable relatively easy: `MiniPairs.bs()` can take a `key` variable used as base key (`<BS>` by default). Then it is a matter for users to make these mappings:

```lua
vim.keymap.set('i', '<C-h>', 'v:lua.MiniPairs.bs()', {expr = true, replace_keycodes = false})
vim.keymap.set('i', '<C-w>', 'v:lua.MiniPairs.bs("\23")', { expr = true, replace_keycodes = false })
vim.keymap.set('i', '<C-u>', 'v:lua.MiniPairs.bs("\21")', { expr = true, replace_keycodes = false })
```

Does this work for you?

------

I think literally mapping `<C-h>` to `<BS>` in Insert mode should be fine. Be sure to use `{ remap = true }`. If you don't like this approach, you can recreate mapping for `<BS>` itself as I've written before.

**Comment by jakubkaczor** [comment]
I believe the idea as you put it is correct. I don't think it acknowledges for the case where the bracket is a letter that can be inside a word, see “sniffing” above, but perhaps it is an edge case, and it would be premature to try to solve for it, if it doesn't raise in practice.

Furthermore, I believe the solution as you described it would work, but the code snippet doesn't. I put it into my config, and then while in the state `(test|)`, pressing `<C-w>` results in `(tes|)`, that is it works like `<BS>` where it should work as before.

Also, with regard to mapping `<C-h>` to `<BS>`, I dug out the reason it could be crippling. It prevents you from mapping `<BS>` to something else, than deleting the character. Actually, in NeoVim I don't use backspace at all, and I could map it to something else, as `<C-h>` is sufficient. Though, for now, I simply don't use it for anything, so it would be a solution. Despite, remapping this key doesn't seem unlikely to happen. `<BS>` is a truly valid key for mapping, as it is unused, but as I said, not the big deal at the moment.

**Comment by echasnovski** [comment]
> Furthermore, I believe the solution as you described it would work, but the code snippet doesn't. I put it into my config, and then while in the state `(test|)`, pressing `<C-w>` results in `(tes|)`, that is it works like `<BS>` where it should work as before.

Yes, it needs small changes to `MiniPairs.bs()` to take key as argument and use it. I plan to add this tomorrow. If you want to pre-test it, update [these two lines](https://github.com/echasnovski/mini.nvim/blob/0c2210fe2e9e08e4dbf33af6984e48a6cc3e0b57/lua/mini/pairs.lua#L328-L329) in your local installed version of 'mini.nvim'/'mini.pairs' to be:

```lua
MiniPairs.bs = function(key)
  local res = key or H.keys.bs
```

> Also, with regard to mapping `<C-h>` to `<BS>`, I dug out the reason it could be crippling. It prevents you from mapping `<BS>` to something else than deleting the character. Actually, in NeoVim I don't use backspace at all, and I could map it to something else, as `<C-h>` is sufficient. Though, for now, I simply don't use it for anything, so it would be a solution. Despite, remapping this key doesn't seem unlikely to happen. `<BS>` is a truly valid key for mapping, as it is unused, but as I said, not the big deal at a moment.

Again, you can map `<C-h>` to use `v:lua.MiniPairs.bs()` as in snippet above. It should work now even without previously mentioned code change.

**Comment by echasnovski** [comment]
Thanks for suggestion. The proposed approach to handling `<C-h>`, `<C-w>`, and `<C-u>` is now in `main` branch. See [this snippet in `bs()` help](https://github.com/echasnovski/mini.nvim/blob/c16a18124da94fd6d35669a2208b78da0fdd18e2/doc/mini-pairs.txt#L283-L289) for how to set them up.

**Comment by jakubkaczor** [comment]
I tested it, and it works, apart from one case that emerged which I didn't mention before. That is having `This (is|) an example.`, pressing `<C-u>` results in `|) an example.`, but `| an example.` is expected. Without looking into code, I assume that the plugin performs a one-character neighborhood check, as removing the bracket in situations like
```
{|
    function();
}
```
also doesn't work this way with any key. Therefore, I consider this outside the scope of this issue, perhaps this plugin. Anyway, the current one has been solved, so thank you for the discussion and quick addressing. 😉
